File: resources/documentation/bdd-testing/how-to-write-bdd-scenarios-cucumber/js/features/step_definitions/login.steps.js
const { Given, When, Then, Before, After } = require('@cucumber/cucumber');
const assert = require('node:assert/strict');

class LoginPage {
  static USERS = { 'jane@qapractices.com': 'SecurePass123!' };

  constructor() {
    this.navigateTo();
  }

  navigateTo() {
    this.email = '';
    this.password = '';
    this.message = '';
    this.onDashboard = false;
  }

  clickButton(name) {
    if (name.toLowerCase() !== 'login') return;
    if (!this.email) {
      this.message = 'Email is required';
      return;
    }
    if (!(this.email in LoginPage.USERS)) {
      this.message = 'User not found';
      return;
    }
    if (LoginPage.USERS[this.email] !== this.password) {
      this.message = 'Invalid credentials';
      return;
    }
    this.message = 'Welcome';
    this.onDashboard = true;
  }

  loginWithSso() {
    this.message = 'Welcome';
    this.onDashboard = true;
  }
}

Before(function () {
  this.page = new LoginPage();
});

After(function (scenario) {
  if (scenario.result.status === 'FAILED') {
    this.attach(`Scenario failed: ${scenario.pickle.name}`, 'text/plain');
  }
});

Given('I am on the login page', function () {
  this.page.navigateTo();
});

When('I enter {string} in the email field', function (email) {
  this.page.email = email;
});

When('I enter {string} in the password field', function (password) {
  this.page.password = password;
});

When('I click the {string} button', function (buttonName) {
  this.page.clickButton(buttonName);
});

When('I log in with SSO', function () {
  this.page.loginWithSso();
});

Then('I should see {string}', function (expectedMessage) {
  assert.equal(this.page.message, expectedMessage);
});

Then('I should be redirected to the dashboard', function () {
  assert.equal(this.page.onDashboard, true);
});

Then('I should see the dashboard', function () {
  assert.equal(this.page.onDashboard, true);
});

File: resources/documentation/bdd-testing/how-to-write-bdd-scenarios-cucumber/behave/features/steps/login_steps.py
from behave import given, when, then


class LoginPage:
    USERS = {"jane@qapractices.com": "SecurePass123!"}

    def __init__(self):
        self.email = ""
        self.password = ""
        self.message = ""
        self.on_dashboard = False

    def navigate_to(self):
        self.email = ""
        self.password = ""
        self.message = ""
        self.on_dashboard = False

    def click_button(self, name):
        if name.lower() != "login":
            return
        if not self.email:
            self.message = "Email is required"
            return
        if self.email not in self.USERS:
            self.message = "User not found"
            return
        if self.USERS[self.email] != self.password:
            self.message = "Invalid credentials"
            return
        self.message = "Welcome"
        self.on_dashboard = True

    def login_with_sso(self):
        self.message = "Welcome"
        self.on_dashboard = True


@given('I am on the login page')
def step_on_login_page(context):
    context.page = LoginPage()
    context.page.navigate_to()


@when('I enter "{email}" in the email field')
def step_enter_email(context, email):
    context.page.email = email


@when('I enter "{password}" in the password field')
def step_enter_password(context, password):
    context.page.password = password


@when('I click the "{button_name}" button')
def step_click_button(context, button_name):
    context.page.click_button(button_name)


@when('I log in with SSO')
def step_login_sso(context):
    context.page.login_with_sso()


@then('I should see "{expected_message}"')
def step_verify_message(context, expected_message):
    assert context.page.message == expected_message


@then('I should be redirected to the dashboard')
def step_verify_dashboard(context):
    assert context.page.on_dashboard


@then('I should see the dashboard')
def step_see_dashboard(context):
    assert context.page.on_dashboard
